gap/attr.gi

```diff
@@ -128,18 +135,29 @@ function(digraph)
     if not IsNullDigraph(subdigraph)
         and not IsBipartiteDigraph(subdigraph) then
       if IsCompleteDigraph(DigraphSymmetricClosure(subdigraph)) then
-        if n > chrom then
+        if chrom = fail or n > chrom then
           chrom := n;
         fi;
       else
-        for i in [3 .. n - 1] do
-          if DigraphColoring(subdigraph, i) <> fail then
-            if i > chrom then
-              chrom := i;
+        upper := RankOfTransformation(DigraphColouring(subdigraph), nr);
+        if chrom = fail or upper > chrom then
+          upper := Minimum(upper, nr - 1);
+          lower := Maximum(chrom, CliqueNumber(subdigraph));
+          if upper = lower then
+            # This line is hard to test since DigraphColouring is
+            # non-deterministic now
+            chrom := upper; # upper > chrom
+          else # lower < upper
+            chrom := upper - 1;
+            repeat
+              coloring := DigraphColouring(subdigraph, chrom);
+              chrom    := chrom - 1;
+            until coloring = fail or chrom = lower;
```

Comment:
Once I fix the other bugs (changing `nr` to `n`, and the initialisation bug with `lower` being equal to `fail`, being greater than any integer), this example highlights the bug I'm on about:

```
gr := DigraphFromGraph6String("KmKk~K??G@_@");
ChromaticNumber(gr);
DigraphColouring(gr, 3);
DigraphColouring(gr, 4);
```
The chromatic number is 4, but we're getting the answer 3. That's because the clique number is 3, which is our lower bound, and we've found using the homomorphisms code that we are able to colour it with 4 colours. Your code is then assuming that that means that it is 3-colourable.

It should be simple to fix.